### PR TITLE
[#9382] Enable generatedOptionsFor in submission page for MSQ/MCQ

### PR DESCRIPTION
**Pull request by xpdavid**
Part of #9382 

**Outline of Solution**

- Logic will inject dynamically generated options into questions before sending to the frontend when indent === SUBMISSION.
- This will also abstract the logic when the MSQ/MCQ question is validated (corresponding tests have been added).
- Besides, this PR also tidies up the MSQ/MCQ answer/edit form in the frontend (e.g., consistent ngModelChange handling and unidirectional data flow (#9660).).

### CODE REVIEW

File: src/web/app/components/question-types/question-edit-details-form/msq-question-edit-details-form.component.ts [hunk 1]
```diff
@@ -103,51 +110,87 @@ export class MsqQuestionEditDetailsFormComponent
...
+        generateOptionsFor: FeedbackParticipantType.NONE,
+        msqChoices: ['', ''],
+        otherEnabled: false,
+        hasAssignedWeights: false,
+        msqWeights: [],
+        msqOtherWeight: 0,
+        minSelectableChoices: NO_VALUE,
+        maxSelectableChoices: NO_VALUE,
+      });
+    }
```
Comment (damithc) [inline src/web/app/components/question-types/question-edit-details-form/msq-question-edit-details-form.component.ts]: How about this? easier to see what is similar and what is different.
```js
  triggerGeneratedOptionsChange(checked: boolean): void {
   foo = checked ? FeedbackParticipantType.STUDENTS : FeedbackParticipantType.NONE
   bar = checked ? ... : ...
      this.triggerModelChangeBatch({
        generateOptionsFor: foo,
        msqChoices: bar
        otherEnabled: false,
        hasAssignedWeights: false,
        msqWeights: [],
        msqOtherWeight: 0,
        minSelectableChoices: NO_VALUE,
        maxSelectableChoices: NO_VALUE,
      });
```

File: src/web/app/components/question-types/question-edit-details-form/msq-question-edit-details-form.component.ts [hunk 2]
```diff
@@ -103,51 +110,87 @@ export class MsqQuestionEditDetailsFormComponent
...
+  triggerMaxSelectableOptionsChange(checked: boolean): void {
+    if (checked) {
+      if (this.isMinSelectableChoicesEnabled) {
+        this.triggerModelChange('maxSelectableChoices', this.model.minSelectableChoices);
+      } else {
+        this.triggerModelChange('maxSelectableChoices', 2);
+      }
+    } else {
+      this.triggerModelChange('maxSelectableChoices', NO_VALUE);
+    }
```
Comment (damithc) [inline src/web/app/components/question-types/question-edit-details-form/msq-question-edit-details-form.component.ts]: Use guard clauses to reduce nesting?

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 3]
```diff
@@ -498,6 +501,102 @@ private boolean areQuestionNumbersConsistent(List<FeedbackQuestionAttributes> qu
...
+                    (FeedbackMcqQuestionDetails) feedbackQuestionAttributes.getQuestionDetails();
+            optionList = feedbackMcqQuestionDetails.getMcqChoices();
+            generateOptionsFor = feedbackMcqQuestionDetails.getGenerateOptionsFor();
+        } else if (feedbackQuestionAttributes.getQuestionType() == FeedbackQuestionType.MSQ) {
+            FeedbackMsqQuestionDetails feedbackMsqQuestionDetails =
+                    (FeedbackMsqQuestionDetails) feedbackQuestionAttributes.getQuestionDetails();
+            optionList = feedbackMsqQuestionDetails.getMsqChoices();
+            generateOptionsFor = feedbackMsqQuestionDetails.getGenerateOptionsFor();
+        } else {
+            return;
```
Comment (damithc) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Is this path expected to be taken? If not, it should be an assertion?

**Comment by xpdavid** [comment]
@damithc Thanks for the suggestions. The code becomes much clearer.

**Pull request merged**